src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,150 @@
+        {
+            var arrayCreationExpression = (IArrayCreationOperation)context.Operation;
+
+            // Don't offer if the user doesn't want it
+            var option = context.GetOption(CSharpCodeStyleOptions.PreferUTF8StringLiteral);
+            if (!option.Value)
+                return;
+
+            // Only replace arrays with initializers
+            if (arrayCreationExpression.Initializer is null)
+                return;
+
+            var byteType = context.Compilation.GetSpecialType(SpecialType.System_Byte);
+            var elementType = (arrayCreationExpression.Type as IArrayTypeSymbol)?.ElementType;
+            if (!SymbolEqualityComparer.Default.Equals(elementType, byteType))
+                return;
+
+            // UTF8 strings are not valid to use in attributes
+            if (arrayCreationExpression.Syntax.Ancestors().OfType<AttributeSyntax>().Any())
+                return;
+
+            // Can't use a UTF8 string inside an expression tree.
+            var semanticModel = context.Operation.SemanticModel;
+            Contract.ThrowIfNull(semanticModel);
+            if (arrayCreationExpression.Syntax.IsInExpressionTree(semanticModel, expressionType, context.CancellationToken))
+                return;
+
+            var values = arrayCreationExpression.Initializer.ElementValues.SelectAsArray(v => v.ConstantValue.HasValue && v.ConstantValue.Value is not null, v => v.ConstantValue.Value);
+
+            // If we couldn't get constant values for all elements then we can't offer
+            if (values.Length != arrayCreationExpression.Initializer.ElementValues.Length)
+                return;
+
+            var stringValue = GetStringValue(values);
+
+            if (stringValue is null)
+                return;
+
+            var properties = ImmutableDictionary<string, string?>.Empty.Add(StringValuePropertyName, stringValue);
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, arrayCreationExpression.Syntax.GetLocation(), option.Notification.Severity, additionalLocations: null, properties));
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        private static string? GetStringValue(ImmutableArray<object?> values)
+        {
+            try
+            {
+                var byteValues = new byte[values.Length];
+                for (var i = 0; i < values.Length; i++)
+                {
+                    // We shouldn't get nulls, but Convert.ToByte will return (byte)0 if there are any
+                    // so better to crash if the analyzer has a bug, that output a buggy string in users code.
+                    if (values[i] is null)
+                        throw ExceptionUtilities.Unreachable;
+
+                    byteValues[i] = Convert.ToByte(values[i]);
```

Comment:
instead of this, actually just validate that everything was a byte constant in the original array.

Reply:
But given `new byte[] { 65,66,67 }` they are ints

Reply:
really?  that's pretty surprising.  there aren't conversion nodes or anything here that have the values as bytes?

Reply:
really?  that's pretty surprising.  there aren't conversion nodes or anything here that have the values as bytes?

Reply:
Ahh you know what, I've still got the `WalkDownConversion` there, from when I was validating syntax being a literal. Will try again.

Reply:
Yeah, this works fine. Removed ToByte, and will move string creation to the code fix.